src/getPublicKeys.c

```diff
@@ -0,0 +1,265 @@
+
+	UI_STEP_BEGIN(ctx->ui_step);
+	UI_STEP(GET_KEY_UI_STEP_WARNING) {
+		ui_displayPaginatedText(
+		        "Unusual request",
+		        "Proceed with care",
+		        this_fn
+		);
+	}
+	UI_STEP(GET_KEY_UI_STEP_DISPLAY) {
+		ui_displayAccountScreen("Export public key", &ctx->pathSpec, this_fn);
+	}
+	UI_STEP(GET_KEY_UI_STEP_CONFIRM) {
+		ui_displayPrompt(
+		        "Confirm export",
+		        "public key?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(GET_KEY_UI_STEP_RESPOND) {
+		ASSERT(ctx->responseReadyMagic == RESPONSE_READY_MAGIC);
+
+		io_send_buf(SUCCESS, (uint8_t*) &ctx->extPubKey, SIZEOF(ctx->extPubKey));
+		ctx->responseReadyMagic = 0; // just for safety
+
+		if (ctx->remainingPaths > 0) {
+			ui_displayBusy(); // waiting for another APDU
+		} else {
+			ui_idle(); // we are done, display the main app screen
+		}
+	}
+	UI_STEP_END(GET_KEY_UI_STEP_INVALID);
+}
+
+// derive the key described by ctx->pathSpec and run the ui state machine accordingly
+// if isSingleKeyExport is true, more restrictive security policy is applied
+static void returnOneKey(bool isSingleKeyExport)
+{
+	ctx->responseReadyMagic = 0;
+
+	// Check security policy
+	security_policy_t policy = (isSingleKeyExport) ?
+	                           policyForGetExtendedPublicKey(&ctx->pathSpec) :
+	                           policyForGetExtendedPublicKeyBulkExport(&ctx->pathSpec);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// Calculation
+		deriveExtendedPublicKey(
+		        & ctx->pathSpec,
+		        & ctx->extPubKey
+		);
+		ctx->responseReadyMagic = RESPONSE_READY_MAGIC;
+	}
+
+	switch (policy) {
+#	define  CASE(policy, step) case policy: {ctx->ui_step = step; break;}
+		CASE(POLICY_PROMPT_WARN_UNUSUAL,    GET_KEY_UI_STEP_WARNING);
+		CASE(POLICY_PROMPT_BEFORE_RESPONSE, GET_KEY_UI_STEP_DISPLAY);
+		CASE(POLICY_ALLOW_WITHOUT_PROMPT,   GET_KEY_UI_STEP_RESPOND);
+#	undef   CASE
+	default:
+		ASSERT(false);
+	}
+	getPublicKeys_getOneKey_ui_runStep();
+}
+
+// ============================== INIT ==============================
+
+enum {
+	HANDLE_INIT_UI_STEP_CONFIRM = 100,
+	HANDLE_INIT_UI_STEP_RESPOND,
+	HANDLE_INIT_UI_STEP_INVALID,
+} ;
+
+static void getPublicKeys_handleInit_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = getPublicKeys_handleInit_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+	UI_STEP(HANDLE_INIT_UI_STEP_CONFIRM) {
+		char secondLine[100];
+		explicit_bzero(secondLine, SIZEOF(secondLine));
+		snprintf(secondLine, SIZEOF(secondLine), "%u public keys?", ctx->remainingPaths);
+		ASSERT(strlen(secondLine) < SIZEOF(secondLine));
+
+		ui_displayPrompt(
+		        "Confirm export",
+		        secondLine,
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_INIT_UI_STEP_RESPOND) {
+		ASSERT(ctx->remainingPaths > 0);
+		returnOneKey(false); // runs another UI state machine
+
+		ctx->stage = GET_KEYS_STAGE_GET_KEYS;
```

Comment:
I am worried about this. returnOneKey is async and I don't know what are the consequences of setting this afterwards ...